pkg/build/strategies/layered/layered.go

```diff
@@ -138,24 +145,33 @@ func (builder *Layered) SourceTar(config *api.Config) (io.ReadCloser, error) {
 // Build handles the `docker build` equivalent execution, returning the
 // success/failure details.
 func (builder *Layered) Build(config *api.Config) (*api.Result, error) {
+	buildResult := &api.Result{}
+
 	if config.HasOnBuild && config.BlockOnBuild {
-		return nil, fmt.Errorf("builder image uses ONBUILD instructions but ONBUILD is not allowed")
+		buildResult.BuildInfo = updater.UpdateFailureReason(buildResult.BuildInfo, api.ReasonGenericS2IBuildFailed)
+		return buildResult, fmt.Errorf("builder image uses ONBUILD instructions but ONBUILD is not allowed")
 	}
 
 	if err := builder.CreateDockerfile(config); err != nil {
-		return nil, err
+		// NOTE: check if CreateDockerfile needs to be added to failure cases.
+		buildResult.BuildInfo = updater.UpdateFailureReason(buildResult.BuildInfo, api.ReasonGenericS2IBuildFailed)
```

Comment:
I'm only seeing CreateDockerfile called with layered and onbuild.  I think it would help to be more specific than generic s2i build failed.   Perhaps even have a unique reason for layered and another for onbuild.